sstables/mx/partition_reversing_data_source.cc

```diff
@@ -0,0 +1,588 @@
+/*
+ * Copyright (C) 2021-present ScyllaDB
+ */
+
+/*
+ * This file is part of Scylla.
+ *
+ * Scylla is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Scylla is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Scylla.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <seastar/core/coroutine.hh>
+#include <seastar/core/iostream.hh>
+#include "partition_reversing_data_source.hh"
+#include "reader_permit.hh"
+#include "sstables/consumer.hh"
+#include "sstables/processing_result_generator.hh"
+#include "sstables/shared_sstable.hh"
+#include "sstables/sstables.hh"
+#include "sstables/types.hh"
+#include "vint-serialization.hh"
+
+namespace sstables {
+
+namespace mx {
+
+// Parser for the partition header and the static row, if present.
+//
+// After consuming the input stream, allows reading the file offset after the consumed segment
+// using header_end_pos()
+// Parsing copied from the sstable reader, with verification removed.
+//
+class partition_header_context : public data_consumer::continuous_data_consumer<partition_header_context> {
+    uint64_t _header_end_pos;
+    bool _ended = false;
+    processing_result_generator _gen;
+    temporary_buffer<char>* _processing_data;
+public:
+    bool non_consuming() const {
+        return false;
+    }
+    void verify_end_state() const {
+        if (!_ended) {
+            throw std::runtime_error("partition_header_context - no more data but parsing is incomplete");
+        }
+    }
+    uint64_t header_end_pos() {
+        return _header_end_pos;
+    }
+    data_consumer::processing_result process_state(temporary_buffer<char>& data) {
+        _processing_data = &data;
+        auto ret = _gen.generate();
+        if (ret == data_consumer::proceed::no) {
+            _ended = true;
+        }
+        return ret;
+    }
+private:
+    uint64_t current_position() {
+        return position() - _processing_data->size();
+    }
+    processing_result_generator do_process_state() {
+        co_yield read_16(*_processing_data);
+        // skip the partition key and the deletion_time
+        co_yield skip(*_processing_data, uint32_t{_u16} + sizeof(uint32_t) + sizeof(uint64_t));
+        co_yield read_8(*_processing_data);
```

Comment:
We need better documentation of the field names in comments (the first read_16 is not documented), otherwise the reader has to reverse-engineer the code to understand it.